st2common/st2common/transport/consumers.py

```diff
@@ -47,26 +47,42 @@ def get_consumers(self, Consumer, channel):
 
     def process(self, body, message):
         try:
+            if not isinstance(body, self._handler.message_type):
+                raise TypeError('Received an unexpected type "%s" for payload.' % type(body))
+
             self._dispatcher.dispatch(self._process_message, body)
-        finally:
-            message.ack()
+        except:
+            LOG.exception('%s failed to process message: %s', self.__class__.__name__, body)
+
+        message.ack()
 
     def _process_message(self, body):
         try:
+            self._handler.process(body)
+        except:
+            LOG.exception('%s failed to process message: %s', self.__class__.__name__, body)
+
+
+class StagedQueueConsumer(QueueConsumer):
+
+    def process(self, body, message):
+        try:
             if not isinstance(body, self._handler.message_type):
                 raise TypeError('Received an unexpected type "%s" for payload.' % type(body))
-
-            self._handler.process(body)
+            processed_body = self._handler.pre_ack_process(body)
+            self._dispatcher.dispatch(self._process_message, processed_body)
         except:
             LOG.exception('%s failed to process message: %s', self.__class__.__name__, body)
+        # At this point we will always ack a specific message.
+        message.ack()
```

Comment:
I really think you should use finally clause. Otherwise someone could accidentally add a return in except clause causing all sorts of damage. Same for above class.